src/algorithms/synthesis/dd_synthesis.cpp

```diff
@@ -433,21 +426,83 @@ namespace syrec {
         return unifyPath(src, current, p1SigVec, p2SigVec, changePaths, dd);
     }
 
-    // This function returns the operations required to synthesize the DD.
-    auto DDSynthesizer::synthesize(dd::mEdge src, std::unique_ptr<dd::Package<>>& dd) -> qc::QuantumComputation& {
-        qc.clear();
-        runtime  = 0.;
-        numGates = 0U;
+    // Refer to the decoder algorithm of https://www.cda.cit.tum.de/files/eda/2018_aspdac_coding_techniques_in_synthesis.pdf.
+    auto DDSynthesizer::decoder(TruthTable::CubeMap const& codeword, std::size_t const& r, std::size_t const& m, std::unique_ptr<dd::Package<>>& dd) -> void {
+        const auto codeLength = codeword.begin()->second.size();
+        const auto totalBits  = qc.getNqubits();
+        // if r = 0, skip to decoder part 2.
+
+        //Decoder part 1.
+        if (r > 0U) {
+            for (const auto& [pattern, code]: codeword) {
+                TruthTable::Cube targetCube(pattern.begin(), pattern.begin() + static_cast<int>(r));
+
+                dd::Controls ctrl;
+                for (auto i = 0U; i < codeLength; i++) {
+                    if (code[i].has_value()) {
+                        const auto ctrlType = *code[i] ? dd::Control::Type::pos : dd::Control::Type::neg;
+                        ctrl.emplace(dd::Control{static_cast<dd::Qubit>((codeLength - 1U) - i), ctrlType});
+                    }
+                }
+
+                const auto targetSize = targetCube.size();
+
+                for (std::size_t i = 0U; i < targetSize; ++i) {
+                    if (targetCube[i].has_value() && *(targetCube[i])) {
+                        auto op = std::make_unique<qc::StandardOperation>(totalBits, ctrl, static_cast<dd::Qubit>((static_cast<std::size_t>(totalBits) - 1U) - i), qc::X);
+                        qc.emplace_back(op);
+                        ++numGates;
+                    }
+                }
+            }
+        }
+
+        //Decoder part 2.
+        const auto correctionBits = m - r;
+
+        if (correctionBits == 0U) {
+            return;
+        }
+
+        TruthTable ttCorrection{};
+
+        for (const auto& [pattern, code]: codeword) {
+            TruthTable::Cube outCube(pattern);
+            outCube.resize(totalBits);
+            TruthTable::Cube inCube(pattern.begin(), pattern.begin() + static_cast<int>(r));
+            for (auto i = 0U; i < codeLength; i++) {
+                inCube.emplace_back(code[i]);
+            }
+            ttCorrection.try_emplace(inCube, outCube);
+        }
+
+        // Extend the dc in the inputs.
+        TruthTable newTT{};
+        for (auto const& [input, output]: ttCorrection) {
+            auto completeInputs = input.completeCubes();
+            for (auto const& completeInput: completeInputs) {
+                newTT.try_emplace(completeInput, output);
+            }
+        }
 
+        const auto ttCorrectionDD = buildDD(newTT, dd);
+        synthesizeDD(ttCorrectionDD, dd, m, true);
+    }
```

Comment:
A rather general thought that crossed my mind: Right now, there is no method that just takes an arbitrary truth table and synthesises it. Such a function would be rather useful/desirable in general as a user-facing method, but could also be used here. I am talking about a function with a signature like
```c++
auto synthesize(const Tableau& tableau) -> qc::QuantumComputation
```
or
```c++
auto synthesize(const Tableau& tableau, qc::QuantumComputation& qc) -> void.
```

In a similar fashion, due to the changes in this PR, there is no function anymore, that simply synthesises a given decision diagram, i.e.,
```c++
auto synthesize(dd::mEdge src, std::unique_ptr<dd::Package<>>& dd) -> qc::QuantumComputation
```
or
```c++
auto synthesize(dd::mEdge src, std::unique_ptr<dd::Package<>>& dd, qc::QuantumComputation& qc) -> void.
```

Reply:
I understood this. I have a question regarding the signature `auto synthesize(const Tableau& tableau) -> qc::QuantumComputation` . I should also consider std::unique_ptr<dd::Package<>>& dd as an argument right?

Suppose instead of using the edge based function, I want to synthesize directly using the reversible truth table itself. For benchmarks like "urf3" and "urf4", I have observed that the garbage nodes start getting collected. So in the test  I use `const auto&   qcDD = dd::buildFunctionality(&qc, dd);`. However, the dd used in the tests is not the same as dd created inside the `auto synthesize(const Tableau& tableau) -> qc::QuantumComputation` function. As a result, `EXPECT_EQ(ttDD, qcDD);` turnsout to be false.

Reply:
No, the general purpose synthesis function does not need a DD package as an argument. It should construct one inside the function. Otherwise, this functionality will be really hard to export to Python since there is no Python interface for the DD package (yet). Furthermore, a potential user won't care about decision diagrams at all. They are just used for turning the TT into a circuit.

If garbage nodes start to be collected, then something in the reference counting is not working as intended and should be looked into.

Reply:
I think you adjusted the reference counting last time right? Any idea what the issue might be?

Reply:
Ah. I think I missed the point above. Two different DD packages will not produce the same DD (they use separate memory and thus the pointer addresses are different even though the structure might be the same).
However, if you use the same package for constructing the original `ttDD` and the construction of the circuit functionality `qcDD` that should yield the same decision diagram.
It might be a bit trickier when the original TT contains don't cares, because you will not get that particular DD from the circuit, but rather one realisation with particular values for the don't cares.

Reply:
yes this makes sense to me. Now the question is how do I use the same package for constructing ` ttDD` and `qcDD`? 
For that to happen, Should I return the same dd package used to construct `ttDD` and use the returned value to construct `qcDD`?

Reply:
you have the original truth table available. the `buildDD` function for creating the DD for a given truth table takes a DD package instance; so does the `buildFunctionality` function. Just use the same package instance for both of these calls in the tests.

Reply:
I think I did not mention my question properly. The above works for the function `auto synthesize(dd::mEdge src, std::unique_ptr<dd::Package<>>& dd, qc::QuantumComputation& qc) -> void.`. What if I want to use the function `auto synthesize(const Tableau& tableau) -> qc::QuantumComputation`? `buildDD` is called inside this function, meaning, the dd is also constructed inside this function. That means I cannot use this in the tests.

Reply:
So right now I'm testing like this,

`EXPECT_TRUE(readPla(tt, fileName));
    extend(tt);

    const auto ttDD = buildDD(tt, dd);
    EXPECT_TRUE(ttDD.p != nullptr);
    DDSynthesizer synthesizer(tt);

    const auto qc = synthesizer.synthesize(tt);
    const auto&   qcDD = dd::buildFunctionality(qc.get(), dd);
    EXPECT_EQ(ttDD, qcDD);`

I assume this is what you are asking me to do. The issue is the `synthesizer.synthesize(tt)` function uses its own package to synthesize and load the gates in qc. 

Now when we reach ` const auto&   qcDD = dd::buildFunctionality(qc.get(), dd);`

the dd here is different (structurally same).